### PR TITLE
[Lib] Fix disconnected lib

### PR DESCRIPTION
**Pull request by adityaramteke**
Lib fix for the changes required for disconnected configuration

### CODE REVIEW

File: cephci/utils/configure.py [hunk 1]
```diff
@@ -277,15 +296,24 @@ def add_images_to_private_registry(node, username, password, registry, build_typ
...
+    con = get_cephci_config()
+    src_username = con.get("cdn_credentials", {}).get("username")
+    src_password = con.get("cdn_credentials", {}).get("password")
+    src_creds = f"{src_username}:{src_password}"
     dest_creds = f"{username}:{password}"
     dst_creds_dir = "./certs/"
 
-    # Get the required images here
-    images = get_images(build_type)
+    # # Get the required images here
```
Comment (vamahaja) [inline cephci/utils/configure.py:307]: ????

File: cli/utilities/containers.py [hunk 2]
```diff
@@ -53,72 +53,85 @@ def __init__(self, nodes, package="podman"):
...
+        """Pull container image
+
+        Args:
+            image (str): Container image url
+        """
+
+        cmd = f"{self.base_cmd} pull"
+        if image:
+            cmd += f" {image}"
+        else:
```
Comment (vamahaja) [inline cli/utilities/containers.py:133]: This else part is not required as name is mandatory parameter

File: cli/utilities/containers.py [hunk 3]
```diff
@@ -53,72 +53,85 @@ def __init__(self, nodes, package="podman"):
...
             raise ContainerRegistryError(f"Failed to run {cmd} on container")
+
+    def pull(self, image):
+        """Pull container image
+
+        Args:
+            image (str): Container image url
+        """
+
+        cmd = f"{self.base_cmd} pull"
```
Comment (vamahaja) [inline cli/utilities/containers.py:130]: this should be `cmd = f"{self.base_cmd} pull name"`

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cephci/prereq.py [hunk 4]
```diff
@@ -32,12 +32,13 @@
...
 from cli.utilities.waiter import WaitUntil
 from utility.log import Log
+from utility.utils import get_cephci_config
 
 log = Log(__name__)
 
 CEPHADM_ANSIBLE = "cephadm-ansible"
-REGISTRY_USERNAME = "Test"
-REGISTRY_PASSWORD = "Test"
+REGISTRY_USERNAME = "myregistryusername"
```
[the remark this answers is not in the document]
Comment (adityaramteke) [reply cephci/prereq.py:40]: Better to have any name than simply saying test with starting upper case letter.

File: cli/utilities/containers.py (hunk 2 above)
Comment (adityaramteke) [reply cli/utilities/containers.py:133]: Sure, I'll remove that.

File: cephci/prereq.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vamahaja) [reply cephci/prereq.py:40]: Default passwords can be anything. Either provide an option to pass these details to library or revert these changes.

File: cephci/prereq.py [hunk 5]
```diff
@@ -201,27 +204,32 @@ def setup_private_container_registry(installer, nodes, registry, build_type):
...
-    # Verify trust list is updated
-    if not validate_trusted_list(nodes[1]):
-        return False
+    # Login to registry
+    con = get_cephci_config()
+    username = con.get("cdn_credentials", {}).get("username")
+    password = con.get("cdn_credentials", {}).get("password")
+    for node in nodes:
+        Registry(node).login(
+            registry="registry.redhat.io", username=username, password=password
```
Comment (vamahaja) [inline cephci/prereq.py:216]: Why hard coded registry ? Will it work for downstream, stage or ibm builds ?

File: cephci/utils/configure.py [hunk 6]
```diff
@@ -277,15 +296,25 @@ def add_images_to_private_registry(node, username, password, registry, build_typ
         f"{DOMAIN_CERT_PATH_CERT}:/certs/domain.cert:Z",
     ]
 
-    src_creds = f"{username}:{password}"
+    # Get login credentials
+    con = get_cephci_config()
+    src_username = con.get("cdn_credentials", {}).get("username")
+    src_password = con.get("cdn_credentials", {}).get("password")
```
Comment (vamahaja) [inline cephci/utils/configure.py:302]: Same as above, this is restricting library to live server. Pass all these details through function parameter

File: cephci/utils/configure.py [hunk 7]
```diff
@@ -277,15 +296,25 @@ def add_images_to_private_registry(node, username, password, registry, build_typ
...
+    src_creds = f"{src_username}:{src_password}"
     dest_creds = f"{username}:{password}"
     dst_creds_dir = "./certs/"
 
     # Get the required images here
-    images = get_images(build_type)
+    images = list(images)
+    if images and type(images) not in (tuple, list):
+        images = [images]
+    else:
```
Comment (vamahaja) [inline cephci/utils/configure.py:311]: Remove this one, keep image as mandatory parameter.

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cephci/utils/configure.py (hunk 6 above)
Comment (adityaramteke) [reply cephci/utils/configure.py:302]: Sure, I'll pass these through a test case.

File: cli/utilities/containers.py [hunk 8]
```diff
@@ -53,72 +53,81 @@ def __init__(self, nodes, package="podman"):
...
+            cmd += f" --name {name}"
+
         if restart:
             cmd += f" --restart={restart}"
 
-        if rm:
-            cmd += f" --rm {rm}"
-        else:
-            cmd += f" {image}"  # Add image only if not an rm operation
+        if name:
```
Comment (vamahaja) [inline cli/utilities/containers.py:89]: repeated `name` parameter check

File: cephci/utils/configure.py [hunk 9]
```diff
@@ -225,19 +226,36 @@ def validate_trusted_list(node):
...
+    """Copy the certificate to another nodes in the cluster"""
+    for secondary_node in secondary_nodes:
+        cmd = f"scp {DOMAIN_CERT_PATH} root@{secondary_node.hostname}:{CA_TRUST_ANCHORS_PATH}"
+        out, _ = node.exec_command(cmd=cmd, sudo=True, timeout=800)
+        if out:
+            log.error(f"Failed to copy the certificate to {secondary_node} node")
+            return False
+
+        # Update the ca-trust
+        if not update_ca_trust(node):
```
Comment (vamahaja) [inline cephci/utils/configure.py:239]: Add log error

File: cephci/utils/configure.py [hunk 10]
```diff
@@ -225,19 +226,36 @@ def validate_trusted_list(node):
...
+            return False
+
+        # Validate the trusted list
+        if not validate_trusted_list(node):
```
Comment (vamahaja) [inline cephci/utils/configure.py:243]: Add log error

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/utilities/containers.py [hunk 11]
```diff
@@ -53,72 +53,78 @@ def __init__(self, nodes, package="podman"):
...
             volume (list): List of volumes
             ports (list): List of ports
             rm (list): List of entries to be removed
-            detach (bool): Detach flag
+            detach (str): Detach flag
             restart (str): Restart flag
             cmds (str): Other commands to be executed
         """
-        if not image and not rm:
+        if not name and not rm:
```
Comment (vamahaja) [inline cli/utilities/containers.py:78]: Why this change ? which image is mandatory, not name

Comment (adityaramteke) [reply cli/utilities/containers.py:78]: Yes, you are correct, based on the general run command, IMAGE is a mandatory parameter:
``` podman run [options] IMAGE [COMMAND [ARG...]] ```

I've some queries regarding options we are currently using for podman run command for disconnected installation:

We are passing image with different parameters for e.g.  --detach IMAGE (Basically to run the container in background) and --rm IMAGE (To remove container when stopped)

 How should we handle such scenarios with the current implementation?

Comment (vamahaja) [reply cli/utilities/containers.py:78]: @adityaramteke as per documentation [1]. 

`name` is not a mandatory parameter. `image` is mandatory and `rm` is optional parameter which requires `image` to be passed.
Also in case  `detach` parameter doesn't have any value then it works as `-d` (which will run container in background) but here `detach` is passed as list or string as `--detach` it accepts multiple string, here we `IMAGE` is used because it's passed at the end.

[1] https://docs.podman.io/en/latest/markdown/podman-run.1.html

Comment (adityaramteke) [reply cli/utilities/containers.py:78]: Thanks, Vaibhav! I've made the changes.

File: cli/utilities/containers.py [hunk 12]
```diff
@@ -53,72 +53,79 @@ def __init__(self, nodes, package="podman"):
...
-            rm (list): List of entries to be removed
-            detach (bool): Detach flag
+            rm (str): Image name to remove from background
+            detach (str): Image name to detach
             restart (str): Restart flag
             cmds (str): Other commands to be executed
         """
-        if not image and not rm:
-            raise NotSupportedError("Image name or rm option needs to be provided")
+        if not image:
```
Comment (vamahaja) [inline cli/utilities/containers.py:78]: This will never be executed.

Comment (adityaramteke) [reply cli/utilities/containers.py]: Resolved.

**Review by vamahaja — COMMENTED** [review]
LGTM


@pranavprakash20 please review

**Review by pranavprakash20 — APPROVED** [review]
LGTM

**Review by vamahaja — APPROVED** [review]
